pkg/security/ptracer/network_handlers.go

```diff
@@ -274,3 +275,78 @@ func shouldSendAccept(msg *ebpfless.SyscallMsg) bool {
 func shouldSendBind(msg *ebpfless.SyscallMsg) bool {
 	return msg.Retval >= 0 || msg.Retval == -int64(syscall.EACCES) || msg.Retval == -int64(syscall.EPERM) || msg.Retval == -int64(syscall.EADDRINUSE) || msg.Retval == -int64(syscall.EFAULT)
 }
+
+func readRemote(pid int, addr uintptr, dst []byte) error {
+	local := []unix.Iovec{{Base: &dst[0], Len: uint64(len(dst))}}
+	remote := []unix.RemoteIovec{{Base: addr, Len: len(dst)}}
+	n, err := unix.ProcessVMReadv(pid, local, remote, 0)
+	if err != nil {
+		return err
+	}
+	if n != len(dst) {
+		return errors.New("short read from remote process")
+	}
+	return nil
+}
+
+func handleSetsockopt(tracer *Tracer, process *Process, msg *ebpfless.SyscallMsg, regs syscall.PtraceRegs, _ bool) error {
+	fd := int32(tracer.ReadArgUint32(regs, 0))
+	socketInfo, ok := process.FdToSocket[fd]
+	if !ok {
+		return errors.New("unable to find socket")
+	}
+
+	level := tracer.ReadArgUint32(regs, 1)
+	optname := tracer.ReadArgUint32(regs, 2)
+
+	// Read the sock_fprog header from argument 3 (optval)
+	// Expected size on 64-bit: 16 bytes (2 bytes length, 6 bytes padding, 8 bytes pointer)
+	hdr, err := tracer.ReadArgData(process.Pid, regs, 3, 16)
+	if err != nil {
+		return err
+	}
+	fLen := binary.NativeEndian.Uint16(hdr[0:2])
+
+	// Pointer at offset 8..15 (due to padding)
+	ptrFilter := binary.NativeEndian.Uint64(hdr[8:16])
+
+	// Nothing to do if no filters
+	if fLen == 0 || ptrFilter == 0 {
+		msg.Type = ebpfless.SyscallTypeSetsockopt
+		msg.Setsockopt = &ebpfless.SetsockoptSyscallMsg{
+			Level:          level,
+			OptName:        optname,
+			Filter:         nil,
+			FilterLen:      0,
+			SocketFamily:   socketInfo.AddressFamily,
+			SocketProtocol: socketInfo.Protocol,
+			SocketType:     socketInfo.SocketType,
+		}
+		return nil
+	}
+
+	const insnSize = 8
+	total := int(fLen) * insnSize
+	raw := make([]byte, total)
+
+	// Read the sock_filter array from the target process's memory
+	if err := readRemote(process.Pid, uintptr(ptrFilter), raw); err != nil {
+		return err
+	}
+
+	msg.Type = ebpfless.SyscallTypeSetsockopt
+	msg.Setsockopt = &ebpfless.SetsockoptSyscallMsg{
+		Level:          level,
+		OptName:        optname,
+		Filter:         raw,
+		FilterLen:      fLen,
+		SocketFamily:   socketInfo.AddressFamily,
+		SocketProtocol: socketInfo.Protocol,
+		SocketType:     socketInfo.SocketType,
+	}
+	return nil
+}
+
+func shouldSendSetsockopt(msg *ebpfless.SyscallMsg) bool {
+	return msg.Retval >= 0 || msg.Retval == -int64(syscall.EACCES) || msg.Retval == -int64(syscall.EPERM) || msg.Retval == -int64(syscall.EINVAL)
```

Comment:
Not sure we want to send the msg when syscall returns EINVAL. If not, we may use the generic `isAcceptedRetval` func.